File: AndroidTesting/Java/EspressoTesting/app/src/androidTest/java/com/dodemy/espressotesting/MainActivityTest.java
package com.dodemy.espressotesting;

import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
import static androidx.test.espresso.action.ViewActions.typeText;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.withHint;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

//Note the static imports, which enhance the code clarity by reducing code length
//import static androidx.test.espresso.Espresso.onView;
//import static androidx.test.espresso.action.ViewActions.click;
//import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
//import static androidx.test.espresso.action.ViewActions.typeText;
//import static androidx.test.espresso.matcher.ViewMatchers.withHint;
//import static androidx.test.espresso.matcher.ViewMatchers.withId;
//import static androidx.test.espresso.assertion.ViewAssertions.matches;
//import static androidx.test.espresso.matcher.ViewMatchers.withText;
//import androidx.test.espresso.ViewAssertion;
//import androidx.test.espresso.action.ViewActions;
//import androidx.test.filters.LargeTest;
//import androidx.test.rule.ActivityTestRule;
//import androidx.test.runner.AndroidJUnit4;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class MainActivityTest {
    // To launch the mentioned activity under testing
    @Rule
    public ActivityScenarioRule<MainActivity> mMainActivityActivityScenarioRule = new ActivityScenarioRule<MainActivity>(MainActivity.class);
    //public ActivityTestRule<MainActivity> mActivityRule = new ActivityTestRule<>(MainActivity.class);

    @Test
    public void testHintVisibility() {
        // check hint visibility
        onView(withId(R.id.editTextName)).check(matches(withHint("Enter Name")));
        // enter name
        onView(withId(R.id.editTextName)).perform(typeText("Pavneet"), closeSoftKeyboard());
        onView(withId(R.id.editTextName)).check(matches(withText("Pavneet")));
    }

    @Test
    public void testButtonClick() {
        // enter name`
        onView(withId(R.id.editTextName)).perform(typeText("Pavneet"), closeSoftKeyboard());
        // clear text
        onView(withText("Clear")).perform(click());
        // check hint visibility after the text is cleared
        onView(withId(R.id.editTextName)).check(matches(withHint("Enter Name")));
    }
}

File: AndroidTesting/Kotlin/EspressoUITest/SimpleActivity/app/src/androidTest/java/com/dodemy/simpleactivity/ui/MainActivityTest.kt
package com.dodemy.simpleactivity.ui

import androidx.test.core.app.ActivityScenario
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import com.dodemy.simpleactivity.R
import org.junit.Test
import org.junit.runner.RunWith


/**
 * Note that the Tests runs in alphabetical order
 * Note that ActivityScenario is used locally for each test, however it can also be
 * declared globally as ActivityScenarioRule with '@get: Rule' annotation
 */

/**
 * @author Maduabughichi Achilefu
 */

@RunWith(AndroidJUnit4ClassRunner::class)
class MainActivityTest {
    //Tests if activity is in view
    @Test
    fun test_isActivityInView() {
        //ActivityScenario mimics an Activity which helps to run each @Test without which the test won't run
        val activityScenario = ActivityScenario.launch(MainActivity::class.java)
        onView(withId(R.id.main))
                .check(matches(isDisplayed()))
    }

    @Test
    fun test_visibilityTitleNextButton() {
        val activityScenario = ActivityScenario.launch(MainActivity::class.java)
        //Tests View visibility
        onView(withId(R.id.activity_main_title))
                .check(matches(isDisplayed())) //Use method type 1
        //Tests View visibility
        onView(withId(R.id.button_next_activity))
                .check(matches(withEffectiveVisibility(Visibility.VISIBLE))) //Or use method type 2
    }

    @Test
    fun test_isTitleTextDisplayed() {
        //ActivityScenario mimics an Activity which helps to run each @Test without which the test won't run
        val activityScenario = ActivityScenario.launch(MainActivity::class.java)
        //Tests visibility of Text Title
        onView(withId(R.id.activity_main_title))
                .check(matches(withText(R.string.text_mainactivity)))
    }

    //Tests navigation to the next activity
    @Test
    fun test_navSecondaryActivity() {
        //Fakes or simulates activity with espresso
        val activityScenario = ActivityScenario.launch(MainActivity::class.java)
        onView(withId(R.id.button_next_activity)).perform(click())
        //Tests to see if the activity is in view/display
        onView(withId(R.id.secondary)).check(matches(isDisplayed()))
    }
}